materialize-snowflake/snowflake.go

```diff
@@ -552,6 +611,108 @@ func (d *transactor) Acknowledge(ctx context.Context) (*pf.ConnectorState, error
 		d.deleteFiles(ctx, []string{item.StagedDir})
 	}
 
+	// Keep asking for a report on the files that have been submitted for processing
+	// until they have all been successful, or an error has been thrown
+	var maxTries = 10
+	for pipeName, pipe := range pipes {
+		for tries := 0; tries < maxTries; tries++ {
+			// We first try to check the status of pipes using the REST API's insertReport
+			// The REST API does not wake up the warehouse, hence our preference
+			// however this API is not the most ergonomic and sometimes does not yield
+			// results as expected
+			report, err := d.pipeClient.InsertReport(pipeName, d.pipeMarkers[pipeName])
+			if err != nil {
+				return nil, fmt.Errorf("snowpipe: insertReports: %w", err)
+			}
+
+			if !report.CompleteResult {
+				tries--
+				d.pipeMarkers[pipeName] = ""
+			} else {
+				d.pipeMarkers[pipeName] = report.NextBeginMark
+			}
+
+			// If there are items still in progress, we continue retrying until those items
+			// resolve to another status
+			var hasItemsInProgress = false
+
+			// If the files have already been loaded, when we submit a request to
+			// load those files again, our request will not show up in reports.
+			// One way to find out whether the files were successfully loaded is to check
+			// the COPY_HISTORY to make sure they are there. The COPY_HISTORY is much more
+			// reliable. If they are not there, then something is wrong.
+			if len(report.Files) == 0 {
+				// We try `maxTries` times since it may take some time for the REST API
+				// to reflect the new pipe requests
+				if tries < maxTries-1 {
+					time.Sleep(5 * time.Second)
+					continue
+				}
+
+				log.WithFields(log.Fields{
+					"tries": tries,
+					"pipe":  pipeName,
+				}).Info("snowpipe: no files in report, fetching copy history from warehouse")
+
+				var fileNames = make([]string, len(pipe.files))
+				for i, f := range pipe.files {
+					fileNames[i] = f.Path
+				}
+
+				rows, err := d.copyHistory(ctx, pipe.tableName, fileNames)
+				if err != nil {
+					return nil, err
+				}
+
+				for _, row := range rows {
+					if row.status == "Loaded" {
+						pipe.fileLoaded(row.fileName)
+					} else if row.status == "Load in progress" {
+						hasItemsInProgress = true
+					} else {
+						return nil, fmt.Errorf("unexpected status %q for files in pipe %q: %s", row.status, pipeName, row.firstErrorMessage)
+					}
+				}
+
+				// If items are still in progress, we continue trying to fetch their results
+				if hasItemsInProgress {
+					tries--
+					time.Sleep(2 * time.Second)
+					continue
```

Comment:
I think these means we'll loop around and try the REST API again. Since the REST API will have demonstrated that it isn't going to give us anything by this point, would it make sense to retry querying `COPY_HISTORY` again (after a delay) instead?

Reply:
@williamhbaker we will retry the REST API, and if it doesn't give us any results again, we will try the COPY_HISTORY again. Note that `tries--` will mean that on the next iteration we will still run `COPY_HISTORY` if no files are given